### PR TITLE
refactor(with-state, structured-list, combobox): replace unsafe lifecycle methods

### PR DESCRIPTION
**Pull request by abbeyhrt**
Refs #4902 

This removes and replaces the functionality of the unsafe lifecycle methods from `withState`, StructuredList, and Combobox. 

#### Changelog

**Removed**

- unsafe lifecycle methods 

#### Testing / Reviewing

Make sure that StructuredList, Combobox, and the components that use withState (Dropdown, DatePicker, and Combobox) work as expected

### CODE REVIEW

File: packages/react/src/tools/withState.js [hunk 1]
```diff
@@ -13,8 +13,11 @@ export default class WithState extends React.PureComponent {
     initialState: PropTypes.object,
   };
 
-  UNSAFE_componentWillMount() {
-    this.setState(this.props.initialState);
+  constructor(props) {
+    super(props);
+    this.state = {
+      initialState: this.props.initialState,
```
Comment (asudoh) [inline packages/react/src/tools/withState.js]: Did you mean `this.state = this.props.initialState`?

Comment (abbeyhrt) [reply packages/react/src/tools/withState.js]: Yep, my bad!

**Review by emyarod — APPROVED** [review]
looks good to me

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @abbeyhrt!

**Review by joshblack — APPROVED** [review]

**Pull request merged**